### PR TITLE
Fixes for testing

### PR DESCRIPTION
**Pull request by tobias-urdin**
Testing some changes to fix testing.

### CODE REVIEW

**Comment by tobias-urdin** [comment]
Waiting for this [1], let's see if a updated image helps here.

[1] https://github.com/gnocchixyz/ci-tools/pull/31

File: setup.cfg [hunk 1]
```diff
@@ -84,15 +84,16 @@ prometheus =
...
     sphinx_rtd_theme
     sphinxcontrib-httpdomain
     PyYAML
     Jinja2
     reno>=1.6.2
 test =
-    pifpaf[ceph,gnocchi]>=1.0.1
-    gabbi>=1.37.0
+    pifpaf[ceph,gnocchi]>=1.0.1,<=2.6.0
+    gabbi==1.49.0
```
Comment (jd) [inline setup.cfg:96]: maybe provide a range?

File: tox.ini [hunk 2]
```diff
@@ -56,10 +56,12 @@ commands =
 recreate = True
 setenv = GNOCCHI_VARIANT=test,postgresql,file
 deps = gnocchi[{env:GNOCCHI_VARIANT}]>=4.3,<4.4
-  pifpaf[gnocchi]>=0.13
+  pifpaf[gnocchi]>=0.13,<=2.6.0
```
Comment (jd) [inline tox.ini:59]: I think `<3` might be more accurate.

File: tox.ini [hunk 3]
```diff
@@ -56,10 +56,12 @@ commands =
...
   gnocchiclient>=2.8.0
   xattr!=0.9.4
-commands = pifpaf --env-prefix INDEXER run postgresql {toxinidir}/run-upgrade-tests.sh {posargs}
+commands =
+  pifpaf --env-prefix INDEXER run postgresql
+  {toxinidir}/run-upgrade-tests.sh {posargs}
```
Comment (jd) [inline tox.ini:64]: unrelated?

File: tox.ini [hunk 4]
```diff
@@ -68,16 +70,21 @@ recreate = True
 setenv = GNOCCHI_VARIANT=test,mysql,ceph,ceph_recommended_lib
 deps = gnocchi[{env:GNOCCHI_VARIANT}]>=4.3,<4.4
   gnocchiclient>=2.8.0
-  pifpaf[ceph,gnocchi]>=0.13
+  pifpaf[ceph,gnocchi]>=0.13,<=2.6.0
```
Comment (jd) [inline tox.ini:73]: ditto

File: tox.ini [hunk 5]
```diff
@@ -68,16 +70,21 @@ recreate = True
...
   xattr!=0.9.4
-commands = pifpaf --env-prefix INDEXER run mysql -- pifpaf --env-prefix STORAGE run ceph {toxinidir}/run-upgrade-tests.sh {posargs}
+commands =
+  pifpaf --env-prefix INDEXER run mysql
+  pifpaf --env-prefix STORAGE run ceph
+  {toxinidir}/run-upgrade-tests.sh {posargs}
```
Comment (jd) [inline tox.ini:78]: ditto

File: tox.ini [hunk 6]
```diff
@@ -95,8 +102,10 @@ deps =
     .[test,file,postgresql,doc]
     doc8
 setenv = GNOCCHI_TEST_DEBUG=1
-commands = doc8 --ignore-path doc/source/rest.rst,doc/source/comparison-table.rst doc/source
-           pifpaf -g GNOCCHI_INDEXER_URL run postgresql -- python setup.py build_sphinx -W
+commands =
+  doc8 --ignore-path doc/source/rest.rst,doc/source/comparison-table.rst doc/source
+  pifpaf -g GNOCCHI_INDEXER_URL run postgresql
+  python setup.py build_sphinx -W
```
Comment (jd) [inline tox.ini:108]: now this is even different

**Review by jd — CHANGES_REQUESTED** [review]

**Comment by tobias-urdin** [comment]
Yeah, some random changes there. Annoying since I had everything passing locally some weeks ago but now something else is broken. The upgrade testing works locally but not in CI.

**Comment by chungg** [comment]
might make our lives (marginally) easier, but can we drop py2.7 tests?

**Comment by jd** [comment]
Yeah we should remove Python 2.7 support now so it'd be fine to remove the tests as a first step.

**Comment by Adrian-Turjak** [comment]
Dropping py2 from testing, and then some follow up work to outright drop py2 (and support py3-isms) isn't a bad idea at all.

**Comment by tobias-urdin** [comment]
The upgrade tests are executed now, but fails when it tries to inject data. For some reason the gnocchi-metricd daemon that is running doesn't process the data. If you start another gnocchi-metricd process in another window it will start processing and continue.

File: .testr.conf [hunk 7]
```diff
@@ -1,5 +1,5 @@
 [DEFAULT]
-test_command=${PYTHON:-python} -m subunit.run discover -t .  ${GNOCCHI_TEST_PATH:-gnocchi/tests} $LISTOPT $IDOPTION
+test_command=${PYTHON:-python3} -m subunit.run discover -t .  ${GNOCCHI_TEST_PATH:-gnocchi/tests} $LISTOPT $IDOPTION
```
Comment (jd) [inline .testr.conf:2]: This should not be needed if you only use Python with tox and Python 3.

File: doc/source/contributing.rst [hunk 8]
```diff
@@ -66,16 +66,16 @@ a virtual environment for each test environment, so make sure you are using an
 up to date version of `virtualenv <https://pypi.python.org/pypi/virtualenv>`_.
 
 Different test environments and configurations can be found by running the
-``tox -l`` command. For example, to run tests with Python 2.7, PostgreSQL as
-indexer, and file as storage backend:
+``tox -l`` command. For example, to run tests with PostgreSQL as indexer,
```
Comment (jd) [inline doc/source/contributing.rst:69]:
```suggestion
``tox -l`` command. For example, to run tests with Python 3.7, PostgreSQL as indexer,
```

File: gnocchi/tests/functional/gabbits/resource-type.yaml [hunk 9]
```diff
@@ -768,8 +768,6 @@ tests:
             min_length: 0
             max_length: 255
       status: 400
-      response_strings:
```
Comment (jd) [inline gnocchi/tests/functional/gabbits/resource-type.yaml:771]: This is actually removing a useful test. Did the error reported changed?

File: run-upgrade-tests.sh [hunk 10]
```diff
@@ -1,9 +1,19 @@
...
+  eval $(pifpaf --debug --env-prefix STORAGE run ceph)
+else
+  echo "error: unsupported upgrade type"
+  exit 1
+fi
 
 export GNOCCHI_DATA=$(mktemp -d -t gnocchi.XXXX)
 
-old_version=$(pip freeze | sed -n '/gnocchi==/s/.*==\(.*\)/\1/p')
+old_version=$(pip3 freeze | sed -n '/gnocchi==/s/.*==\(.*\)/\1/p')
```
Comment (jd) [inline run-upgrade-tests.sh:16]: ditto should not be needed

File: run-upgrade-tests.sh [hunk 11]
```diff
@@ -40,7 +50,7 @@ inject_data() {
 
     {
         measures_sep=""
-        MEASURES=$(python -c 'import datetime, random, json; now = datetime.datetime.utcnow(); print(json.dumps([{"timestamp": (now - datetime.timedelta(seconds=i)).isoformat(), "value": random.uniform(-100000, 100000)} for i in range(0, 288000, 10)]))')
+        MEASURES=$(python3 -c 'import datetime, random, json; now = datetime.datetime.utcnow(); print(json.dumps([{"timestamp": (now - datetime.timedelta(seconds=i)).isoformat(), "value": random.uniform(-100000, 100000)} for i in range(0, 288000, 10)]))')
```
Comment (jd) [inline run-upgrade-tests.sh:53]: ditto

File: run-upgrade-tests.sh [hunk 12]
```diff
@@ -72,17 +82,17 @@ else
...
+eval $(pifpaf --debug run gnocchi --indexer-url $INDEXER_URL --storage-url $STORAGE_URL)
 export OS_AUTH_TYPE=gnocchi-basic
 export GNOCCHI_USER=$GNOCCHI_USER_ID
 original_statsd_resource_id=$GNOCCHI_STATSD_RESOURCE_ID
 inject_data $GNOCCHI_DATA
 dump_data $GNOCCHI_DATA/old
 pifpaf_stop
 
-new_version=$(python setup.py --version)
+new_version=$(python3 setup.py --version)
```
Comment (jd) [inline run-upgrade-tests.sh:93]: ditto

**Review by jd — CHANGES_REQUESTED** [review]
a lot of the python -> python3 should not be needed

File: gnocchi/tests/functional/gabbits/resource-type.yaml (hunk 9 above)
Comment (tobias-urdin) [reply gnocchi/tests/functional/gabbits/resource-type.yaml]: Yes, changed the output. Will revert and check output again.

**Review by jd — COMMENTED** [review]
Looks like upgrade fails because it's too verbose :)

**Comment by tobias-urdin** [comment]
Feedback on how to change gabbits response_string if this doesn't work :)

**Comment by jd** [comment]
Gentle ping on @cdent if he knows something changed in Gabbi :)

**Comment by jd** [comment]
I wonder if the upgrade problem is not due to gnocchiclient recent issue?

**Comment by cdent** [comment]
> Gentle ping on @cdent if he knows something changed in Gabbi :)

Can you point me to where things are going wrong?

One thing I did notice is that regex (wrapping strings with `/`) are not available in `response_strings`; it checks the entire response body for the string fragment, so that's a thing that needs to be fixed.

But if you can link me a failure I can provide more.

**Comment by tobias-urdin** [comment]
@cdent https://travis-ci.org/github/gnocchixyz/gnocchi/jobs/725172818#L447

Is there a way to check if the response_body contains the value instead of what it tries there to check the complete value? Seems like the output has changed in that response from the API probably due to some dependency (uwsgi or whatever it might be).

**Comment by cdent** [comment]
> Is there a way to check if the response_body contains the value instead of what it tries there to check the complete value?

Thanks for the link. Yeah, the issue there is that you're trying to use a regex in a `response_strings`. Each entry in the list provided with `response_strings` is a string fragment. You want something like:

```yaml
[...]
response_strings:
  - "can't remove a non-existent object 'what'"
[...]
```

If that doesn't work for some reason it is probably a wart with YAML quoting, in which case, try matching on a smaller string.

The `/foo/` syntax you're trying to use works in a few other places, but not in `response_strings` (since it is a list of fragments)

**Comment by chungg** [comment]
> I wonder if the upgrade problem is not due to gnocchiclient recent issue?

yeah. could be related to https://github.com/gnocchixyz/python-gnocchiclient/pull/110? 

@tobias-urdin you can try forcing it to pip install gnocchiclient==7.0.6 to see if it's related to one of the PRs merged.

File: gnocchi/tests/functional/gabbits/resource-type.yaml [hunk 13]
```diff
@@ -769,7 +769,7 @@ tests:
             max_length: 255
       status: 400
       response_strings:
-       - "can't remove non-existent object 'what'"
+       - can't remove non-existent object 'what'
```
Comment (cdent) [inline gnocchi/tests/functional/gabbits/resource-type.yaml:772]: The word `a` is missing from this message. See the test failure for the exact message, but I think it's supposed to be "can't remove _a_ non-existent..."

File: gnocchi/tests/functional/gabbits/resource-type.yaml [hunk 14]
```diff
@@ -819,7 +819,7 @@ tests:
           path: /attributes/unknown
       status: 400
       response_strings:
-       - "can't remove non-existent object 'unknown'"
+       - can't remove non-existent object 'unknown'
```
Comment (cdent) [inline gnocchi/tests/functional/gabbits/resource-type.yaml:822]: same here:

The word `a` is missing from this message. See the test failure for the exact message, but I think it's supposed to be "can't remove _a_ non-existent..."

**Review by cdent — CHANGES_REQUESTED** [review]

File: gnocchi/tests/functional/gabbits/resource-type.yaml (hunk 13 above)
Comment (chungg) [reply gnocchi/tests/functional/gabbits/resource-type.yaml:772]: eagle eyes.

Comment (cdent) [reply gnocchi/tests/functional/gabbits/resource-type.yaml:772]: I'm a pterodactyl, caw caw.

Comment (jd) [reply gnocchi/tests/functional/gabbits/resource-type.yaml:772]: Says the _eye of lynx_.

http://eavesdrop.openstack.org/irclogs/%23openstack-telemetry/%23openstack-telemetry.2016-12-06.log.html#t2016-12-06T17:04:48

https://en.wiktionary.org/wiki/%C5%93il_de_lynx 😛

Comment (chungg) [reply gnocchi/tests/functional/gabbits/resource-type.yaml:772]: /me looks at date of chat log

how is this possible? i can't remember yesterday.

Comment (tobias-urdin) [reply gnocchi/tests/functional/gabbits/resource-type.yaml]: Wow, crazy how I could have missed that. Let's see if it works when fixing that. Thanks @cdent!

**Comment by tobias-urdin** [comment]
Yay, tests passing :)

**Comment by Adrian-Turjak** [comment]
Fantastic!

File: run-upgrade-tests.sh [hunk 15]
```diff
@@ -40,7 +50,7 @@ inject_data() {
 
     {
         measures_sep=""
-        MEASURES=$(python -c 'import datetime, random, json; now = datetime.datetime.utcnow(); print(json.dumps([{"timestamp": (now - datetime.timedelta(seconds=i)).isoformat(), "value": random.uniform(-100000, 100000)} for i in range(0, 288000, 10)]))')
+        MEASURES=$(python -c 'import datetime, random, json; now = datetime.datetime.utcnow(); print(json.dumps([{"timestamp": (now - datetime.timedelta(seconds=i)).isoformat(), "value": random.uniform(-100000, 100000)} for i in range(0, 1000, 10)]))')
```
Comment (jd) [inline run-upgrade-tests.sh:53]: The range was this big to be sure we tested file splitting IIRC. Might be good to keep it that high.

File: tox.ini [hunk 16]
```diff
@@ -57,27 +58,28 @@ recreate = True
...
 
 [testenv:pep8]
+basepython = python3
 deps = hacking>=0.12,<0.13
 commands = flake8
 
-[testenv:py27-cover]
-commands = pifpaf -g GNOCCHI_INDEXER_URL run postgresql -- python setup.py testr --coverage --testr-args="{posargs}"
+[testenv:py37-cover]
+commands = pifpaf -g GNOCCHI_INDEXER_URL run postgresql -- python3 setup.py testr --coverage --testr-args="{posargs}"
```
Comment (jd) [inline tox.ini:82]: you shouldn't need python3 here

File: run-upgrade-tests.sh [hunk 17]
```diff
@@ -1,6 +1,16 @@
 #!/bin/bash
 set -e
 
+if [ "$1" == "postgresql-file" ]; then
+  eval $(pifpaf --env-prefix INDEXER run postgresql)
+elif [ "$1" == "mysql-ceph" ]; then
+  eval $(pifpaf --env-prefix INDEXER run mysql)
```
Comment (jd) [inline run-upgrade-tests.sh:7]: If you use `eval` here, nothing will stop mysql/postgresql for you.

You'll have to call `pifpaf_stop` IIRC to stop it before you finish this script or the process will kept hanging around.

That's why `pifpaf` was invoked from `tox`.

**Review by jd — CHANGES_REQUESTED** [review]
Almost good. I think there are a few changes that might not be needed, like using `eval`, which will leave hanging processes once the tests are done.

File: run-upgrade-tests.sh (hunk 15 above)
Comment (tobias-urdin) [reply run-upgrade-tests.sh:53]: I will try adding it back see if Travis complains about output length again.

File: run-upgrade-tests.sh (hunk 17 above)
Comment (tobias-urdin) [reply run-upgrade-tests.sh:7]: Let me know if new change is better

File: run-upgrade-tests.sh [hunk 18]
```diff
@@ -61,6 +71,8 @@ pifpaf_stop(){
 cleanup(){
     pifpaf_stop
```
Comment (jd) [inline run-upgrade-tests.sh:72]: You have `pifpaf_stop` here not sure it works if `indexer_stop` works?
I mean there's too many stop begin called now I think :)

**Review by jd — CHANGES_REQUESTED** [review]

File: run-upgrade-tests.sh [hunk 19]
```diff
@@ -61,6 +71,8 @@ pifpaf_stop(){
...
     rm -rf $GNOCCHI_DATA
+    indexer_stop
+    [ -n $STORAGE_DAEMON ] && storage_stop
```
Comment (chungg) [inline run-upgrade-tests.sh:75]: https://github.com/gnocchixyz/gnocchi/blob/master/setup.cfg#L94 seems to be always enabling ceph so STORAGE_DAEMON gets set. you may want to add a clause to ensure it's not file or to remove ceph from setup.cfg (although not sure what other things gets affected)

File: run-upgrade-tests.sh (hunk 18 above)
Comment (tobias-urdin) [reply run-upgrade-tests.sh:72]: I don't think so, the pifpaf_stop should still be the non-prefixed service running under pifpaf which is Gnocchi, while indexer_stop is, well, the indice database :)

File: run-upgrade-tests.sh (hunk 19 above)
Comment (tobias-urdin) [reply run-upgrade-tests.sh:75]: Ah, yeah. Thanks I'll try to do a check instead.

File: run-upgrade-tests.sh [hunk 20]
```diff
@@ -61,6 +71,8 @@ pifpaf_stop(){
 cleanup(){
     pifpaf_stop
     rm -rf $GNOCCHI_DATA
+    indexer_stop
+    [ "$STORAGE_DAEMON" == "ceph" ] && storage_stop
```
Comment (chungg) [inline run-upgrade-tests.sh:75]: just need to add `|| true` so it doesn't return non zero code.

Comment (tobias-urdin) [reply run-upgrade-tests.sh]: Thanks!

**Comment by tobias-urdin** [comment]
Tests are passing again, let me know if there is any additional feedback.

**Review by jd — APPROVED** [review]

**Comment by Adrian-Turjak** [comment]
*cheers* TY Tobias!

I was gonna look at rebasing https://github.com/gnocchixyz/gnocchi/pull/995 but it seems you're already looking at it. :)

**Comment by tobias-urdin** [comment]
@jd can we try to backport this to stable/4.3 shouldn't be any breaking changes

**Comment by tobias-urdin** [comment]
i assume i can't send mergify commands

**Comment by tobias-urdin** [comment]
@Mergifyio backport stable/4.3

**Comment by jd** [comment]
@Mergifyio backport stable/4.3

**Comment by jd** [comment]
@tobias-urdin you should do it manually :(

**Comment by tobias-urdin** [comment]
ok, thanks :)
